feat(templates): new templates for issues and PRs

gauthierpetetin opened the pull request:
**Description**

We've discussed with stakeholders from Extension and Mobile teams how we could align templates for issues and PRs on the different repos, starting with metamask-extension and metamask-mobile repos. [This](https://docs.google.com/document/d/1HMvxQIYUBVOgJ2Iqccn3uZ9Pu003_l1RWIfo1fMHvjE/edit) is the proposal we landed on after our first workshops.

**Screenshots/Recordings**

Templates can be visualised on this test repo:
- [General issue](https://github.com/gauthierpetetin-test/repo_test/issues/new?assignees=&labels=&projects=&template=general_issue.yml)
- [Bug report issue](https://github.com/gauthierpetetin-test/repo_test/issues/new?assignees=&labels=type-bug&projects=&template=bug-report-mobile.yml&title=%5BBug%5D%3A+)
- [Pull request](https://github.com/gauthierpetetin-test/repo_test/pull/126)

**Issue**

NA

**References**

- Same PR on metamask-mobile repo: https://github.com/MetaMask/metamask-mobile/pull/7076

gauthierpetetin committed:
feat(templates): new templates for issues and PRs

We've discussed with stakeholders from Extension and Mobile teams how we could align templates for issues and PRs on the different repos, starting with metamask-extension and metamask-mobile repos. This is the proposal we landed on after our first workshops.

gauthierpetetin committed:
fix(templates): links were not working

DDDDDanica commented on .github/coding_guidelines/LABELING_GUIDELINES.md:
```diff
@@ -0,0 +1,27 @@
+- **needs-qa**: If the PR includes a new features, complex testing steps, or large refactors, this label must be added to indicated PR requires a full manual QA prior being merged and added to a release.
+
+### Labels prohibited when PR needs to be merged:
+Any PR that includes one of the following labels can not be merged:
```
```suggestion
Any PR that includes one of the following labels cannot be merged:
```

gauthierpetetin replied:
Both "cannot" and "can not" are technically correct.
The latter is used to emphasize the "not" part which, was the intention here.

DDDDDanica commented on .github/coding_guidelines/CODING_GUIDELINES.md:
```diff
@@ -0,0 +1,76 @@
+
+### 1. New Code Should be Typescript
+- New components and utilities should be written in Typescript and enforce typing.
+- Existing code should be refactored into Typescript where time allows. If you are replacing a component use Typescript.
```
```suggestion
- Existing code should be refactored into Typescript where time allows. If you are replacing a component, use Typescript.
```

gauthierpetetin replied:
Fixed in this [commit](https://github.com/MetaMask/metamask-extension/pull/20651/commits/725877d0181a3bdf6b965a6c5aab017824b7f728)

DDDDDanica commented on .github/coding_guidelines/CODING_GUIDELINES.md:
```diff
@@ -0,0 +1,76 @@
+AvatarAccount
+├── AvatarAccount.constants.ts
+├── AvatarAccount.stories.tsx
+├── AvatarAccount.styles.ts
```
we are still using scss for styles, i would say this can be `avatar-account.scss`?

gauthierpetetin replied:
I'm wondering what is the long term guideline we want to promote for styles? `scss` or `ts`?
In Mobile `ts` is indicated in the [guideline](https://github.com/MetaMask/metamask-mobile/blob/3f43e586d2e442b66295d1ede01eec49ead91146/.github/coding_guidelines/CODING_GUIDELINES.md?plain=1#L17).

Is there a reason to have a different guideline for Extension and Mobile?

DDDDDanica replied:
Actually i don't have a concrete answer for this, i would say @georgewrmarshall might have better idea about this one !

gauthierpetetin replied:
Hi @NicolasMassart , as I'd like us to have the same coding guidelines on Extension and Mobile side, can I please ask for your recommendation on this question?

gauthierpetetin replied:
Hi @georgewrmarshall @NicolasMassart do you have any recommendation on this topic?

georgewrmarshall replied:
Hey @gauthierpetetin @DDDDDanica, Apologies for the late reply. We use `.ts` in mobile and `.scss` in the extension codebases for distinct reasons relating to how styles are written in different languages.

- In mobile, we are using a CSS-in-JS approach where the styles are written in TypeScript. 
- In extension, we write styles in SCSS which is transpiled into CSS.

We can't consolidate the 2 here.

gauthierpetetin replied:
Ok makes sense, thanks for the detail. It now updated in this [commit](https://github.com/MetaMask/metamask-extension/pull/20651/commits/4b5fdc9c3ee5acfc767c7ca5668badec31c58cb9).

DDDDDanica commented on .github/coding_guidelines/CODING_GUIDELINES.md:
````diff
@@ -0,0 +1,76 @@
+
+```.tsx
+AvatarAccount
+├── AvatarAccount.constants.ts
+├── AvatarAccount.stories.tsx
+├── AvatarAccount.styles.ts
````
```suggestion
├── avatar-account.constants.ts
├── avatar-account.stories.tsx
├── avatar-account.styles.scss
```

gauthierpetetin replied:
cf [comment](https://github.com/MetaMask/metamask-extension/pull/20651#discussion_r1309962231)

DDDDDanica commented on .github/coding_guidelines/CODING_GUIDELINES.md:
```diff
@@ -0,0 +1,76 @@
+- When creating hooks use *withHookName()*.
+
+### 5. Avoid Repetitive Code
+- If you notice you are writing duplicated code or components, convert it into a component, utility functions or hooks that can be reused. Do this with [scalable intention](https://sandimetz.com/blog/2016/1/20/the-wrong-abstraction).
```
👍🏻

DDDDDanica commented on .github/coding_guidelines/CODING_GUIDELINES.md:
```diff
@@ -0,0 +1,76 @@
+- If you notice you are writing duplicated code or components, convert it into a component, utility functions or hooks that can be reused. Do this with [scalable intention](https://sandimetz.com/blog/2016/1/20/the-wrong-abstraction).
+
+### 6. Component Optimization
+- For functional components, instead of having large return statements, break the component down into smaller sub-components.
```
```suggestion
- For functional components, instead of having large return statements, breaking the component down into smaller sub-components.
```

gauthierpetetin replied:
This sentence is in the imperative mood, giving a direct instruction or command to the reader, which is why "break" is preferred to "breaking".

DDDDDanica commented on .github/coding_guidelines/CODING_GUIDELINES.md:
```diff
@@ -0,0 +1,76 @@
+
+### 6. Component Optimization
+- For functional components, instead of having large return statements, break the component down into smaller sub-components.
+- Use memoizing techniques where possible. Utilize the useMemo hook for values and useCallback for functions. Follow recommended React Native guidance.
+
```
```suggestion
- Use memoizing techniques where possible. Utilize the `useMemo` hook for values and `useCallback` for functions. Follow recommended React Native guidance.

```

gauthierpetetin replied:
Fixed in this [commit](https://github.com/MetaMask/metamask-extension/pull/20651/commits/725877d0181a3bdf6b965a6c5aab017824b7f728)

DDDDDanica commented on .github/coding_guidelines/CODING_GUIDELINES.md:
```diff
@@ -0,0 +1,76 @@
+
+### 6. Component Optimization
+- For functional components, instead of having large return statements, break the component down into smaller sub-components.
+- Use memoizing techniques where possible. Utilize the useMemo hook for values and useCallback for functions. Follow recommended React Native guidance.
```
Should we also include `useEffect`, and also indicate not to ignore the dependencies

gauthierpetetin replied:
I was hoping we could have coding guidelines, that apply to both Extension and Mobile.
Would this also apply to Mobile? Can you please indicate the sentence you think we shall add?

DDDDDanica replied:
how about this:
```
`useEffect` for performing data fetching or dom manipulation after the component has rendered 
```

gauthierpetetin replied:
Same here @NicolasMassart , as I'd like us to have the same coding guidelines on Extension and Mobile side, can I please ask for your recommendation on this question?

HowardBraham replied:
I would also encourage us to read and link to https://react.dev/learn/you-might-not-need-an-effect

gauthierpetetin replied:
@DDDDDanica @NicolasMassart @HowardBraham if good for you I will add the following comment to both Extension and Mobile repos coding guidelines later today.

`Use the useEffect hook for performing side effects like data fetching or DOM manipulation after the component has rendered. However, use it judiciously as unnecessary effects can complicate code and affect performance. For a deeper understanding, we recommend reading [this article](https://react.dev/learn/you-might-not-need-an-effect).`

DDDDDanica commented on .github/coding_guidelines/CODING_GUIDELINES.md:
```diff
@@ -0,0 +1,76 @@
+- Use memoizing techniques where possible. Utilize the useMemo hook for values and useCallback for functions. Follow recommended React Native guidance.
+
+### 7. Use Object Destructuring For Props
+- Instead of passing the props object, use object destructuring to pass the prop name. This discards the need to refer to the props object each time you need to use it.
```
```suggestion
- Instead of passing the props object, using object destructuring to pass the prop name. This discards the need to refer to the props object each time you need to use it.
```

gauthierpetetin committed:
fix(templates): tiny adjustments

gauthierpetetin committed:
fix(templates): add expected behavior section to bug report template

This change was suggested by Nicolas Massart.

gauthierpetetin committed:
fix(templates): add a section to indicate what issue is fixed by the PR

This change was suggested by Nicolas Massart.

gauthierpetetin committed:
fix(guidelines): add a coding guideline regarding the use of useEffect

gauthierpetetin committed:
fix(guidelines): rename guidelines folder

gauthierpetetin committed:
fix(contributing.md): update link to good first issues

gauthierpetetin committed:
fix(contributing.md): recommend contributors to read coding guidelines

georgewrmarshall commented on .github/guidelines/CODING_GUIDELINES.md:
```diff
@@ -0,0 +1,77 @@
+# MetaMask Coding Guidelines
+
+### 1. New Code Should be Typescript
```
```suggestion
### 1. New Code Should be TypeScript
```

georgewrmarshall commented on .github/guidelines/CODING_GUIDELINES.md:
```diff
@@ -0,0 +1,77 @@
+# MetaMask Coding Guidelines
+
+### 1. New Code Should be Typescript
+- New components and utilities should be written in Typescript and enforce typing.
```
nit: Update casing here
```suggestion
- New components and utilities should be written in TypeScript and enforce typing.
```

gauthierpetetin replied:
Fixed in this [commit](https://github.com/MetaMask/metamask-extension/pull/20651/commits/57532c023692c4f8650785ab90e29c9f5d8881ca).

georgewrmarshall commented on .github/guidelines/CODING_GUIDELINES.md:
```diff
@@ -0,0 +1,77 @@
+# MetaMask Coding Guidelines
+
+### 1. New Code Should be Typescript
```
nit: Update casing here
```suggestion
### 1. New Code Should be TypeScript
```

gauthierpetetin replied:
Fixed in this [commit](https://github.com/MetaMask/metamask-extension/pull/20651/commits/57532c023692c4f8650785ab90e29c9f5d8881ca).

georgewrmarshall commented on .github/guidelines/CODING_GUIDELINES.md:
```diff
@@ -0,0 +1,77 @@
+
+### 1. New Code Should be Typescript
+- New components and utilities should be written in Typescript and enforce typing.
+- Existing code should be refactored into Typescript where time allows. If you are replacing a component, use Typescript.
```
nit: Update casing here
```suggestion
- Existing code should be refactored into TypeScript where time allows. If you are replacing a component, use TypeScript.
```

gauthierpetetin replied:
Fixed in this [commit](https://github.com/MetaMask/metamask-extension/pull/20651/commits/57532c023692c4f8650785ab90e29c9f5d8881ca).

georgewrmarshall commented on .github/guidelines/CODING_GUIDELINES.md:
````diff
@@ -0,0 +1,77 @@
+- An example of a component file structure:
+
+```.tsx
+AvatarAccount
+├── AvatarAccount.constants.ts
+├── AvatarAccount.stories.tsx
+├── AvatarAccount.styles.ts
+├── AvatarAccount.test.tsx
+├── AvatarAccount.tsx
+├── AvatarAccount.types.ts
+├── README.md
+├── __snapshots__
+│   └── AvatarAccount.test.tsx.snap
+└── index.ts
````
I don't think there is any reason why we couldn't consolidate file name casing here but the current convention is kebab-case. I think this is the same as @DDDDDanica suggestion e.g.
```suggestion
avatar-account
├── avatar-account.stories.tsx
├── avatar-account.scss
├── avatar-account.test.tsx
├── avatar-account.tsx
├── avatar-account.types.ts
├── README.md
├── __snapshots__
│   └── avatar-account.test.tsx.snap
└── index.ts
```

gauthierpetetin replied:
Fixed in this [commit](https://github.com/MetaMask/metamask-extension/pull/20651/commits/4b5fdc9c3ee5acfc767c7ca5668badec31c58cb9).

georgewrmarshall commented on .github/guidelines/CODING_GUIDELINES.md:
````diff
@@ -0,0 +1,77 @@
+```
+
+### 4. Follow Naming Conventions
+- You should always use PascalCase when naming components to differentiate them from other non-component TSX files. For example: *TextField*, *NavMenu*, and *SuccessButton*.
````
Because we use kebab-case for file names but retain PascalCase for component and enum names it may make sense to reword this sentence
```suggestion
- You should always use PascalCase when naming components. For example: *TextField*, *NavMenu*, and *SuccessButton*.
```

gauthierpetetin replied:
Fixed in this [commit](https://github.com/MetaMask/metamask-extension/pull/20651/commits/021be0ca2ee2f31f70a29b0e1a1d493d229a9099).

georgewrmarshall commented on .github/guidelines/CODING_GUIDELINES.md:
````diff
@@ -0,0 +1,77 @@
+- Instead of passing the props object, use object destructuring to pass the prop name. This discards the need to refer to the props object each time you need to use it.
+
+```tsx
+import React from 'react';
+import { View } from 'react-native';
+
+const MyComponent = ({id}) => {
+  return <View id={id} />;
+};
+
+const MyComponent = (props, context) => {
+  const { id } = props;
+  return <View id={id} />;
+};
+
+const Foo = class extends React.PureComponent {
+  render() {
+    const { title } = this.context;
+    return <View>{title}</View>
+  }
+};
+
````
Update to use react web syntax and remove react-native specific code. React context is handled differently and we shouldn't be using PureComponents very often
```suggestion
import React from 'react';

const MyComponent = ({ id }) => {
  return <div id={id} />;
};
```

gauthierpetetin replied:
Fixed in this [commit](https://github.com/MetaMask/metamask-extension/pull/20651/commits/021be0ca2ee2f31f70a29b0e1a1d493d229a9099).

georgewrmarshall commented on .github/guidelines/LABELING_GUIDELINES.md:
```diff
@@ -1,17 +1,20 @@
-### Mandatory Labels:
-- **Internal Developers**: Every PR raised by an internal developer must include a label prefixed with `team-` (e.g., `team-extension-ux`, `team-extension-platform`, etc.). This indicates the respective internal team responsible for the PR.
+### Mandatory team labels:
+- **Internal Developers**: Every PR raised by an internal developer must include a label prefixed with `team-` (e.g., `team-mobile-ux`, `team-mobile-platform`, etc.). This indicates the respective internal team responsible for the PR.
```
```suggestion
- **Internal Developers**: Every PR raised by an internal developer must include a label prefixed with `team-` (e.g., `team-extension-ux`, `team-extension-platform`, etc.). This indicates the respective internal team responsible for the PR.
```

gauthierpetetin replied:
Fixed in this [commit](https://github.com/MetaMask/metamask-extension/pull/20651/commits/021be0ca2ee2f31f70a29b0e1a1d493d229a9099).

georgewrmarshall commented on .github/guidelines/CODING_GUIDELINES.md:
```diff
@@ -0,0 +1,77 @@
+
+### 6. Component Optimization
+- For functional components, instead of having large return statements, breaking the component down into smaller sub-components.
+- Use memoizing techniques where possible. Utilize the `useMemo` hook for values and `useCallback` for functions. Follow recommended React Native guidance.
```
Update reference to React Native to React as this is for web
```suggestion
- Use memoizing techniques where possible. Utilize the `useMemo` hook for values and `useCallback` for functions. Follow recommended React guidance.
```

gauthierpetetin replied:
Fixed in this [commit](https://github.com/MetaMask/metamask-extension/pull/20651/commits/021be0ca2ee2f31f70a29b0e1a1d493d229a9099).

georgewrmarshall commented on .github/guidelines/CODING_GUIDELINES.md:
```diff
@@ -0,0 +1,77 @@
+### 1. New Code Should be Typescript
+- New components and utilities should be written in Typescript and enforce typing.
+- Existing code should be refactored into Typescript where time allows. If you are replacing a component, use Typescript.
+
```
We could reference the contributor docs TypeScript best practices here. Something like
```suggestion
Follow contributor doc [TypeScript Guidelines](https://github.com/MetaMask/contributor-docs/blob/main/docs/typescript.md)

```

gauthierpetetin replied:
Added in this [commit](https://github.com/MetaMask/metamask-extension/pull/20651/commits/147f556dfd97dd3f426d9d9f8bc6a01a2c1f27ff).

georgewrmarshall commented on .github/guidelines/CODING_GUIDELINES.md:
```diff
@@ -0,0 +1,77 @@
+- If applicable add URL to online resources if they are meaningful for the component/method.
+
+### 9. Write Tests for Each Component/Utility
+- Write tests for the components you create as it reduces the possibilities of errors. Testing ensures that the components are behaving as you would expect. In this project Jest is used, and it provides an environment where you can execute your tests.
```
We could add a link to the contributor doc unit testing guidelines here
```suggestion
- Write tests for the components you create as it reduces the possibilities of errors. Testing ensures that the components are behaving as you would expect. In this project Jest is used, and it provides an environment where you can execute your tests. Follow the contributor docs [Unit Testing Guidelines](https://github.com/MetaMask/contributor-docs/blob/main/docs/unit-testing.md)
```

gauthierpetetin replied:
Added in this [commit](https://github.com/MetaMask/metamask-extension/pull/20651/commits/147f556dfd97dd3f426d9d9f8bc6a01a2c1f27ff).

georgewrmarshall reviewed (commented):
Looking good! Some suggestions based on the specifics of the extension and removing references to React Native which is only used in mobile. We could also utilize the contributor docs that the Shared Libraries team is working on. I've suggested a couple places to link out to but there may be some more oppurtunities https://github.com/MetaMask/contributor-docs cc @mcmire @desi @Gudahtt

gauthierpetetin committed:
fix(guidelines): coding guidelines for component file structure are different in extension

gauthierpetetin committed:
fix(guidelines): TypeScript shall be written in camelcase

gauthierpetetin committed:
fix(guidelines): simplify component naming guideline

Comment from George Marshall: Because we use kebab-case for file names but retain PascalCase for component and enum names it may make sense to reword this sentence

gauthierpetetin committed:
fix(guidelines): tiny changes to adapt mobile guidelines to extension context

gauthierpetetin committed:
feat(guidelines): add links to contributor docs

gauthierpetetin committed:
fix(issue template): make instruction clearer for external contributors

gauthierpetetin commented on .github/ISSUE_TEMPLATE/general-issue.yml:
```diff
@@ -0,0 +1,105 @@
+    id: design
+    attributes:
+      label: Design
+      placeholder: Link to Figma design or screenshots
```
Good point, I added more detail for external contributors within this [commit](https://github.com/MetaMask/metamask-extension/pull/20651/commits/4cf13dd6b13a6ab444f6831ea8b1949168372f8e).
